aspects/schema.go

```diff
@@ -0,0 +1,437 @@
+					return err
+				}
+			} else {
+				return fmt.Errorf(`unexpected field %q in map`, key)
+			}
+		}
+	}
+
+	if v.keyType != nil {
+		for k := range level {
+			rawKey, err := json.Marshal(k)
+			if err != nil {
+				return err
+			}
+
+			if err := v.keyType.Validate(rawKey); err != nil {
+				return err
+			}
+
+		}
+	}
+
+	if v.valueType != nil {
+		for _, val := range level {
+			if err := v.valueType.Validate(val); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (v *mapSchema) Parse(raw json.RawMessage) error {
+	var level map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &level); err != nil {
+		return err
+	}
+
+	requiredRaw, ok := level["required"]
+	if ok {
+		if err := json.Unmarshal(requiredRaw, &v.requiredCombs); err != nil {
+			return fmt.Errorf(`cannot unmarshal map's "required" field: %v`, err)
+		}
+	}
+
+	// a map can have a "schema" constrain with specific values
+	if schema, ok := level["schema"]; ok {
+		var nextLevel map[string]json.RawMessage
+		if err := json.Unmarshal(schema, &nextLevel); err != nil {
+			return fmt.Errorf(`cannot unmarshal map's "schema" field: %v`, err)
+		}
+
+		v.entryTypes = make(map[string]Schema, len(nextLevel))
+		for key, value := range nextLevel {
+			validator, err := v.baseSchema.Parse(value)
+			if err != nil {
+				return fmt.Errorf(`cannot parse constraint for key %q: %w`, key, err)
+			}
+
+			v.entryTypes[key] = validator
+		}
+
+		return nil
+	}
+
+	// alternatively, it can constrain the type of its keys and/or values
+	keyDescription, ok := level["keys"]
+	if ok {
+		var err error
+		v.keyType, err = v.baseSchema.Parse(keyDescription)
+		if err != nil {
+			return fmt.Errorf(`cannot parse "keys" constraint in map schema: %w`, err)
+		}
+	}
+
+	valuesDescriptor, ok := level["values"]
+	if ok {
+		var err error
+		v.valueType, err = v.baseSchema.Parse(valuesDescriptor)
+		if err != nil {
+			return fmt.Errorf(`cannot parse "values" constraint in map schema: %w`, err)
+		}
+	}
+
+	return nil
+}
+
+type intSchema struct {
+	min     *int
+	max     *int
+	choices []int
+}
+
+func (v *intSchema) Validate(raw []byte) error {
+	var num int
+	if err := json.Unmarshal(raw, &num); err != nil {
+		return err
+	}
+
+	if v.min != nil && num < *v.min {
+		return fmt.Errorf(`integer %d is less than allowed minimum %d`, num, *v.min)
+	}
+
+	if v.max != nil && num > *v.max {
+		return fmt.Errorf(`integer %d is greater than allowed maximum %d`, num, *v.max)
+	}
+
+	if len(v.choices) != 0 {
+		var found bool
+		for _, choice := range v.choices {
+			if num == choice {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return fmt.Errorf(`integer %d is not one of the allowed choices`, num)
+		}
+	}
+
+	return nil
+}
+
+func (v *intSchema) Parse(raw json.RawMessage) error {
+	constraints, err := getTypeConstraints(raw, "int")
+	if err != nil {
+		return err
+	} else if constraints == nil {
+		return nil
+	}
+
+	if rawChoices, ok := constraints["choices"]; ok {
```

Comment:
the intention is for choices to be mutually exclusive with other constraints, same also above for strings